fix(static-themes): Warn on LWT aliases used in a static theme manifest.

rpl opened the pull request:
This PR introduces a new `MANIFEST_THEME_LWT_ALIAS` rule, which is meant to warn the static themes authors of the incoming deprecation of the LWT alias properties.

In this PR this rule is currently considered a warning, which should be turned into an error once we want to prevent new theme submission from using this LWT aliases. 

Fixes #2259 

TODO lists:

- [x] double check with @nt1m that I didn't missed any static theme LWT aliases that should be marked as deprecated
  - [x] add `toolbar_text` to the deprecated aliases (aliased to `bookmark_text`, see [Bug 1472740 comment 8](https://bugzilla.mozilla.org/show_bug.cgi?id=1472740#c8))
- [x] double check with @eviljeff any additional part that may need to happen on the AMO server side (e.g. as part of the conversion of the LWT themes into static themes, we may want to avoid any of the LWT alias properties that are going to be deprecated) 
- [x] submit mozilla-central patch to add the deprecated property on the theme API schemas (as part of  [Bug 1472740 Remove aliased property names from themes](https://bugzilla.mozilla.org/show_bug.cgi?id=1472740))
- [x] rebase this PR on a recent repository head (so that this PR is rebased on top of the imported API schemas introduced in #2325)
- [x] fix test failure related to the linter warnings generated by the mdn-browser-compat-data validation rules 
- [x] change the short link included in the linting rule message's description property to one that points to a section of an MDN doc page which help the author to apply the needed changes (mdn/sprints#821)
- [x] file a follow up to turn `MANIFEST_THEME_LWT_ALIAS` rule from a warning into an error (#2463)
- [x] file a follow up to handle the `deprecated` keyword for the other API schemas (prevented in this PR to avoid regressions unrelated to the goal of this particular issue) (filed as #2462)

rpl commented on src/const.js:
```diff
@@ -122,6 +122,13 @@ export const MIME_TO_FILE_EXTENSIONS = {
 export const STATIC_THEME_IMAGE_MIMES = Object.keys(MIME_TO_FILE_EXTENSIONS);
 
+// List of deprecated static theme's LWT aliases.
+export const STATIC_THEME_LWT_ALIASES = [
```
@nt1m do you mind to confirm if these are all the static theme manifest properties that has to be marked as deprecated LWT aliases?

Thanks in advance for your help!

rpl commented on src/parsers/manifestjson.js:
```diff
@@ -109,6 +110,19 @@ export default class ManifestJSONParser extends JSONParser {
+        // Overwrite the message with the shorter one included in the linter messages.
+        overrides.message = baseObject.message;
+      }
+      // TODO: add a messages.MANIFEST_FIELD_DEPRECATED and ensure that deprecated
```
I'll file a follow up for this (in this PR I opted to prevent the `deprecated` keyword implementation from marking as error any other manifest property or API marked as deprecated in the schema files, because those `deprecated` schema definition have not been processed until now, and so I think that it would be better to evaluate that part separately in a follow up).

rpl commented on src/parsers/manifestjson.js:
```diff
@@ -136,7 +150,12 @@ export default class ManifestJSONParser extends JSONParser {
+      messages.MANIFEST_PERMISSIONS.code,
+      // Remove the following once the LWT aliases deprecated property should
+      // become errors on submission.
+      messages.MANIFEST_THEME_LWT_ALIAS.code,
```
Once we are ready to mark these LWT aliases as errors instead of warnings, on the linter side we just need to remove this line (and so I'm going to file a follow up issue).

EnTeQuAk replied:
Mind adding it to the comment if filed?

rpl commented on src/parsers/manifestjson.js:
```diff
@@ -149,7 +168,10 @@ export default class ManifestJSONParser extends JSONParser {
-      log.debug('Schema Validation messages', validate.errors);
+      log.debug(
+        'Schema Validation messages',
+        JSON.stringify(validate.errors, null, 2)
```
Not strictly related to this PR, but while working on it I noticed that the error were logged as the string '[Object]'... which is not very helpful as a debug log message :-)

rpl commented on src/schema/imported/theme.json:
```diff
@@ -194,7 +194,15 @@
+                  "$ref": "manifest#/types/ImageDataOrExtensionURL"
+                },
+                {
+                  "$comment": "This part has been imported manually and it is expected to be overwritten once the changes to the Firefox schemas reach beta",
```
I imported this manually from a local mozilla-central tree, so that this PR is complete and it can be reviewed, and also because this way we can start to the "deprecated LWT aliases" warning messages even before the change on the mozilla-central side are going to reach beta (and I added some jsonschema comments so that we are going to be aware that these are meant to be overwritten on the next schema updates).

EnTeQuAk replied:
Potentially file a "Import Firefox 65 schema" issue and make a note about that there so that whoever does it, reads it, I'm not sure I'd look for `$comment` in the schema tbh.

rpl replied:
I already rebased this PR on top of the Firefox 65 imported schema, and so this has been resolved.

rpl commented on src/schema/validator.js:
```diff
@@ -55,6 +56,27 @@ validator.addFormat(
 
+validator.addKeyword('deprecated', {
+  validate: function validateDeprecated(message, propValue, schema, dataPath) {
+    if (!STATIC_THEME_LWT_ALIASES.includes(dataPath)) {
```
At line 36 of this file, we are defining a "deprecated" format, but `deprecated` is actually a keyword and so it looks that we never actually took into account any of the "deprecated" keywords used in the schema files.

To avoid regressions related to new unexpected deprecated warnings/errors messages, this PR is currently ignoring any deprecation that is not coming from the static theme LWT aliases's data paths.

I'll file a follow up so that we can discuss and evaluate in more detail how to allow the deprecated keyword to work on the rest of the API schemas.

rpl replied:
I filed #2462 to follow up on this.

eviljeff commented:
> double check with @eviljeff any additional part that may need to happen on the AMO server side

We'd need an issue - see https://github.com/mozilla/addons-linter/issues/2259#issuecomment-440618994

> (e.g. as part of the conversion of the LWT themes into static themes, we may want to avoid any of the LWT alias properties that are going to be deprecated)

The LWT migration project is too far along to incorporate these kind of changes now, so they'll have to picked up afterwards.

EnTeQuAk commented on src/messages/manifestjson.js:
```diff
@@ -296,6 +296,17 @@ export function corruptIconFile({ path }) {
+  // TODO: update the short link included in the description with one that
+  // points directly to a section which describes the changes needed.
+  description: i18n._(
+    'See https://mzl.la/1ZOhoEN (MDN Docs) for more information.'
```
And make that link key something named like `mzl.la/mdn-lwt-alias-deprecation` ideally but that's optional, really.

rpl replied:
@EnTeQuAk I've replaced the link with one that points to https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/theme#Aliases, unfortunately it seems that if I try to customize the link then it becomes a bit.ly link instead of a mzl.la link (e.g. https://mzl.la/2T11Lkc is a mzl.la link that points to the above url, if I create a customized one it becomes http://bit.ly/mdn-lwt-alias-deprecation but http://mzl.la/mdn-lwt-alias-deprecation doesn't exist :-( ).

As additional side note, the additional docs about the deprecation are not in that section yet, the MDN issue for those docs is mdn/sprints#821.

EnTeQuAk replied:
Ah, that's quite unfortunate. really hard, I don't know which is better, generally a customized link is more user friendly as a user kinda knows where he's getting to and to be very fair, mzl.la does safeguard the user somewhat but I don't really know if anyone notices :man_shrugging:

rpl replied:
> Ah, that's quite unfortunate. really hard, I don't know which is better, generally a customized link is more user friendly as a user kinda knows where he's getting to and to be very fair, mzl.la does safeguard the user somewhat but I don't really know if anyone notices man_shrugging

Yeah, it is unfortunate :-(
I ended to prefer the mzl.la one for consistency with the urls included in the other messages defined in `src/messages/manifestjson.js`.

EnTeQuAk commented on src/parsers/manifestjson.js:
```diff
@@ -193,6 +194,19 @@ export default class ManifestJSONParser extends JSONParser {
+      // TODO(#2462): add a messages.MANIFEST_FIELD_DEPRECATED and ensure that deprecated
+      // properties are handled properly (e.g. we should also detect when the deprecated
+      // keyword is actually used to warn the developer of additional properties not
+      // explicitly defined in the schemas).
```
tbh, if we have an issue filed we might as well just move that comment there and remove it here. No one really will read it in the code anyway I think.

rpl commented:
@EnTeQuAk I added a reference to #2463 as part of the inline comment related to turn MANIFEST_THEME_LWT_ALIAS into an error, and rebased the entire PR on top of the current head of the master branch.

rpl commented:
@EnTeQuAk it looks that the docs related to the warning are now part of the MDN doc page section linked from the warning message: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/theme#Aliases

This should now be good to go! \o/
